src/XrdMacaroons/XrdMacaroonsHandler.cc

```diff
@@ -109,19 +145,167 @@ Handler::GenerateActivities(const XrdHttpExtReq & req) const
+}
+
+
+int Handler::ProcessOAuthConfig(XrdHttpExtReq &req) {
+    if (req.verb != "GET")
+    {
+        return req.SendSimpleResp(400, NULL, NULL, "Only GET is valid for oauth config.", 0);
+    }
+    auto header = req.headers.find("Host");
+    if (header == req.headers.end())
+    {
+        return req.SendSimpleResp(400, NULL, NULL, "Host header is required.", 0);
+    }
+
+    json_object *response_obj = json_object_new_object();
+    if (!response_obj)
+    {
+        return req.SendSimpleResp(500, NULL, NULL, "Unable to create new JSON response object.", 0);
+    }
+    std::string token_endpoint = "https://" + header->second + "/.oauth2/token";
+    json_object *endpoint_obj =
+        json_object_new_string_len(token_endpoint.c_str(), token_endpoint.size());
+    if (!endpoint_obj)
+    {
+        return req.SendSimpleResp(500, NULL, NULL, "Unable to create a new JSON macaroon string.", 0);
+    }
+    json_object_object_add(response_obj, "token_endpoint", endpoint_obj);
+
+    const char *response_result = json_object_to_json_string_ext(response_obj, JSON_C_TO_STRING_PRETTY);
+    int retval = req.SendSimpleResp(200, NULL, NULL, response_result, 0);
+    json_object_put(response_obj);
+    return retval;
+}
+
+
+int Handler::ProcessTokenRequest(XrdHttpExtReq &req)
+{
+    if (req.verb != "POST")
+    {
+        return req.SendSimpleResp(400, NULL, NULL, "Only POST is valid for token request.", 0);
+    }
+    auto header = req.headers.find("Content-Type");
+    if (header == req.headers.end())
+    {
+        return req.SendSimpleResp(400, NULL, NULL, "Content-Type missing; not a valid macaroon request?", 0);
+    }
+    if (header->second != "application/x-www-form-urlencoded")
+    {
+        return req.SendSimpleResp(400, NULL, NULL, "Content-Type must be set to `application/macaroon-request' to request a macaroon", 0);
+    }
+    char *request_data;
+    if (req.BuffgetData(req.length, &request_data, true) != req.length)
+    {
+        return req.SendSimpleResp(400, NULL, NULL, "Missing or invalid body of request.", 0);
+    }
+    bool found_grant_type = false;
+    ssize_t validity = -1;
+    std::string scope;
+    std::string token;
+    std::istringstream token_stream(request_data);
+    while (std::getline(token_stream, token, '&'))
+    {
+        std::string::size_type eq = token.find("=");
+        if (eq == std::string::npos)
+        {
+            return req.SendSimpleResp(400, NULL, NULL, "Invalid format for form-encoding", 0);
+        }
+        std::string key = token.substr(0, eq);
+        std::string value = token.substr(eq + 1);
+        //std::cout << "Found key " << key << ", value " << value << std::endl;
+        if (key == "grant_type")
+        {
+            found_grant_type = true;
+            if (value != "client_credentials")
+            {
+                return req.SendSimpleResp(400, NULL, NULL, "Invalid grant type specified.", 0);
+            }
+        }
+        else if (key == "expire_in")
```

Comment:
this "expire_in" grant parameter (which I guess is used to limit the access token lifetime) is something where we want to be aligned, or FTS will have to juggle. StoRM uses the 'lifetime' parameter. No problem to change on our side, but I suggest to have a profile document that defines these details somewhere and align implementations to such profile.

Reply:
Yes - this was a bit arbitrary on my part.  I didn't know of an "obvious" candidate here.

Any reason that `expires_in` versus `expire_in` versus `lifetime` might be preferred?  Shall we start flipping coins?